src/gui/editors/PianoRoll.cpp

```diff
@@ -4803,6 +4834,118 @@ void PianoRollWindow::ghostPatternSet( bool state )
 
 
 
+void PianoRollWindow::exportPattern()
+{
+	QString extFilter("XML Pattern (*.xpt *.xptz)");
+	FileDialog exportDialog(this, tr("Export pattern"), "", extFilter);
+
+	exportDialog.setAcceptMode(FileDialog::AcceptSave);
+
+	if (exportDialog.exec() == QDialog::Accepted &&
+		!exportDialog.selectedFiles().isEmpty() &&
+		!exportDialog.selectedFiles().first().isEmpty())
+	{
+		QString suffix =
+			ConfigManager::inst()->value("app", "nommpz").toInt() == 0
+				? "xptz"
+				: "xpt";
+		exportDialog.setDefaultSuffix(suffix);
+
+		QString fullPath = exportDialog.selectedFiles()[0];
+		QString chosenSuffix = fullPath.section( '.', -1 );
+
+		bool compress = (chosenSuffix == "xpt") ? false : true;
+
+		if (!savePatternXML(fullPath, compress))
+		{
+			TextFloat::displayMessage(tr("Export pattern failed"),
+				tr("No permission to write to %1").arg(fullPath),
+				embed::getIconPixmap("error"), 4000);
+			return;
+		}
+		TextFloat::displayMessage(tr("Export pattern success"),
+			tr("Pattern saved to %1").arg(fullPath),
+			embed::getIconPixmap("project_export"), 4000);
+	}
+}
+
+
+
+
+void PianoRollWindow::importPattern()
+{
+	// Overwrite confirmation.
+	if (!m_editor->m_pattern->empty() &&
+		QMessageBox::warning(
+			NULL,
+			tr("Import pattern."),
+			tr("You are about to import a pattern, this will "
+				"overwrite your current pattern. Do you want to "
+				"continue?"),
+			QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes
+		) != QMessageBox::Yes)
+	{
+		return;
+	}
+
+	FileDialog importDialog(this, tr("Open Pattern"), "",
+		tr("XML pattern file (*.xpt *.xptz)"));
+	importDialog.setFileMode(FileDialog::ExistingFile);
+
+	if (importDialog.exec() == QDialog::Accepted &&
+		!importDialog.selectedFiles().isEmpty())
+	{
+		QString fullPath = importDialog.selectedFiles()[0];
+		QString suffix = fullPath.section('.', -1);
+		bool compression = (suffix == "xpt") ? false : true;
+
+		QDomDocument doc("xml");
+		QFile f(fullPath);
+
+		if (!f.open(QIODevice::ReadOnly))
+		{ // Check if we may read the file.
+			TextFloat::displayMessage(tr("Import pattern failed"),
+				tr("No permission to read file %1").arg(fullPath),
+				embed::getIconPixmap("error"), 4000);
+			f.close();
+			return;
+		}
+
+		bool contentSet = false;
+		if (compression)
+		{
+			contentSet = doc.setContent(qUncompress(f.readAll()));
+		}
+		else
+		{
+			contentSet = doc.setContent(&f);
+		}
+
+		// Check if valid xml.
+		if (!contentSet)
+		{
+			TextFloat::displayMessage(tr("Import pattern failed"),
+				tr("Pattern file corrupt %1").arg(fullPath),
+				embed::getIconPixmap("error"), 4000);
+			f.close();
+			return;
+		}
+		f.close();
+
+		int pos = m_editor->m_pattern->startPosition(); // Backup position in timeline.
```

Comment:
```suggestion
		TimePos pos = m_editor->m_pattern->startPosition(); // Backup position in timeline.
```

I know it works because `TimePos` is converted to `int` without issues, but I think it's better to use the former.